# **AVIATION DATASET DATA ANALYSIS PROJECT**

## **1. Objectives**<hr>

#### **Business Question**<br>
Your company is expanding in to new industries to diversify its portfolio. Specifically, they are interested in purchasing and operating airplanes for commercial and private enterprises, but do not know anything about the potential risks of aircraft. You are charged with determining which aircraft are the lowest risk for the company to start this new business endeavor. You must then translate your findings into actionable insights that the head of the new aviation division can use to help decide which aircraft to purchase.

### Objective 1<br>
Which aircraft is the safest?

### Objective 2<br>
Identify the safest and most dangerous countries & locations to fly over

### Objective 3<br>
The most common type of accidents so as to invest in advance in prevetive measures

<br>

## **2. Data Understanding**<hr>

In this section we try to understand the data itself. First we have to read the source notes on what is in the columns, the categories in categorical and what missing data means. We look at its shape and also at the descriptive statistics of the data.

**Importing libraries and data**

In [1]:
#Importing libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#Importing the data
acc_df = pd.read_csv("data/aviation-accident-data-2023-05-16.csv")

Lets look at the top and bottom 10 rows of the dataset

In [3]:
acc_df.head(10)

date             type registration  \
0  date unk.   Antonov An-12B       T-1206   
1  date unk.   Antonov An-12B       T-1204   
2  date unk.   Antonov An-12B       T-1201   
3  date unk.  Antonov An-12BK          NaN   
4  date unk.  Antonov An-12BP   CCCP-11815   
5  date unk.  Antonov An-12BP   CCCP-12172   
6  date unk.     Antonov An-2    CCCP-N574   
7  date unk.     Antonov An-2   CCCP-01216   
8  date unk.   Antonov An-24B     RA-47794   
9  date unk.    Antonov An-26       01 red   

                               operator fatalities             location  \
0                         Indonesian AF        NaN                  NaN   
1                         Indonesian AF        NaN                  NaN   
2                         Indonesian AF        NaN                  NaN   
3                             Soviet AF        NaN  Tiksi Airport (IKS)   
4                             Soviet AF          0  Massawa Airport ...   
5                             Soviet AF        NaN                  NaN   
6  GUSMP, Directorate of Polar Aviation        NaN              unknown   
7                              Aeroflot          0         Chita region   
8                            Russian AF          0                  NaN   
9                             Soviet AF          0    Orenburg Air Base   

           country cat     year  
0  Unknown country  U1  unknown  
1  Unknown country  U1  unknown  
2  Unknown country  U1  unknown  
3           Russia  A1  unknown  
4          Eritrea  A1  unknown  
5           Russia  U1  unknown  
6           Russia  A1  unknown  
7           Russia  A2  unknown  
8           Russia  A1  unknown  
9           Russia  O1  unknown

In [4]:
acc_df.tail(10)

date                       type registration  \
23957  21-APR-2023           Ilyushin Il-76TD       ST-BDN   
23958  21-APR-2023   Beech 200 Super King Air       LV-WPM   
23959  26-APR-2023           Boeing 777-233LR       C-FIUF   
23960  09-MAY-2023     Cessna 208 Caravan 675       PK-HVG   
23961  10-MAY-2023                Learjet 36A        N56PA   
23962  11-MAY-2023               Hawker 900XP       PK-LRU   
23963  11-MAY-2023  Cessna 208B Grand Caravan       PK-NGA   
23964  12-MAY-2023  Cessna 208B Grand Caravan       5X-RBR   
23965  14-MAY-2023            Boeing 747-4R7F       LX-OCV   
23966  15-MAY-2023                Learjet 35A       D-CGFQ   

                     operator fatalities                  location  \
23957           Badr Airlines          0       Khartoum Interna...   
23958               Flytec SA          0       Salta-Martín Mi...   
23959              Air Canada          0       Sydney-Kingsford...   
23960             Dimonim Air          0             Yabi Airstrip   
23961             Phoenix Air          3  near San Clemente Isl...   
23962  Angkasa Super Services          0       Maleo Airport (MOH)   
23963  Nasional Global Aviasi          0         Fentheik Airstrip   
23964            Bar Aviation          0       Kampala-Kajjansi...   
23965                Cargolux          0       Luxembourg-Finde...   
23966                     GFD          2             Hohn Air Base   

          country cat  year  
23957       Sudan  O1  2023  
23958   Argentina  A2  2023  
23959   Australia  A2  2023  
23960   Indonesia  A2  2023  
23961         USA  A1  2023  
23962   Indonesia  A2  2023  
23963   Indonesia  A2  2023  
23964      Uganda  A1  2023  
23965  Luxembourg  A2  2023  
23966     Germany  A1  2023

At a first glance we get a rough Idea about the format of data in cells. We can also see what null valuesare denoted as especialy in the non-number columns, where they are denoted as unknown

In [5]:
# Getting the shape of data
acc_df.shape

(23967, 9)

The dataframe has 23967 rows and 9 columns. Any column with less than 23967 entries has null values

In [6]:
# Lets look at the column infomation
acc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23967 entries, 0 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          23967 non-null  object
 1   type          23967 non-null  object
 2   registration  22419 non-null  object
 3   operator      23963 non-null  object
 4   fatalities    20029 non-null  object
 5   location      23019 non-null  object
 6   country       23967 non-null  object
 7   cat           23967 non-null  object
 8   year          23967 non-null  object
dtypes: object(9)
memory usage: 1.6+ MB


In [7]:
acc_df["fatalities"].describe()

count     20029
unique      369
top           0
freq      10713
Name: fatalities, dtype: object

We can see all the columns are of object datatype. Fatalities column off the rip needs to be converted to a number. Registration, operator, fatalities and location columns all have null values that have to be dealt with.

<br>

**Descriptive Statistics of the data**

In this section we get the descriptive statistics of the columns

In [8]:
acc_df.describe()

date                  type registration operator fatalities  \
count         23967                 23967        22419    23963      20029   
unique        15079                  3201        21962     6017        369   
top     10-MAY-1940  Douglas C-47A (DC-3)       LZ-...    USAAF          0   
freq            171                  1916           13     2604      10713   

       location country    cat   year  
count     23019   23967  23967  23967  
unique    14608     232     11    106  
top     unknown     USA     A1   1944  
freq        272    4377  17424   1505

Here we can see the descriptive statistics of the non numerical data. We can infer:

    - A1 is the most common accident.
    - 10th May 1940 had alot of accidents
    - Most accidents occured in the year 1944
    - Most accidents had no fatalities.
    - Most accidents occured in the US.
    - The countries might have spelling errors since there are 232 unique countries compared to the 195-200 countries in the world.

In [9]:
# Converting the fatalities  column to int datatype to get correct descriptive statistics
acc_df["fatalities"] = pd.to_numeric(acc_df["fatalities"], errors="coerce") 
#Converts string values to int and if the srting is in a non-int format it isrplaced with NaN

In [10]:
acc_df["fatalities"].describe()

count    19676.000000
mean         6.135851
std         18.686195
min          0.000000
25%          0.000000
50%          0.000000
75%          5.000000
max        520.000000
Name: fatalities, dtype: float64

We can see most accidents were non fatal. Also mean is not a good measure of central tendency here since the data is left-skewed.

<br>

## **3. Data Cleaning**<hr>
In this section we clean the data i.e try to remove the percieved errors in data

First let's handle missing values

In [11]:
# Calling .info on the dataframe again to see where the missing values lie
acc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23967 entries, 0 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   date          23967 non-null  object 
 1   type          23967 non-null  object 
 2   registration  22419 non-null  object 
 3   operator      23963 non-null  object 
 4   fatalities    19676 non-null  float64
 5   location      23019 non-null  object 
 6   country       23967 non-null  object 
 7   cat           23967 non-null  object 
 8   year          23967 non-null  object 
dtypes: float64(1), object(8)
memory usage: 1.6+ MB


Registration won't be important for our endeavour so it can. We will need to handle the facilities data. We will also drop the type of aircrafts that are unknown. 

In [12]:
#Lets take a look at the type first
acc_df["type"].value_counts().to_csv("./output/types.csv")

In [13]:
acc_df[acc_df["type"] == "unknown"]

date     type registration         operator  fatalities  \
5446   20-SEP-1946  unknown          NaN             CNAC        31.0   
5655   25-APR-1947  unknown       ..-141             CNAC         NaN   
6178   30-JAN-1949  unknown          NaN             CNAC         0.0   
6537   11-AUG-1950  unknown          NaN              CSA         0.0   
7036   18-APR-1952  unknown          NaN              JAT         0.0   
8670   08-JUL-1959  unknown          NaN              JAT         0.0   
8904   21-AUG-1960  unknown          NaN         Aeroflot         1.0   
8972   08-DEC-1960  unknown          NaN           Cubana         1.0   
10170  31-DEC-1965  unknown          NaN         Aeroflot         1.0   
10721  19-FEB-1968  unknown          NaN         Aeroflot         2.0   
11210  29-JUL-1969  unknown          NaN          Unknown         0.0   
11537  09-JUN-1970  unknown          NaN              LOT         0.0   
11636  10-SEP-1970  unknown          NaN          Unknown         0.0   
11638  12-SEP-1970  unknown          NaN          Unknown         0.0   
11711  10-DEC-1970  unknown          NaN              CSA         0.0   
11866  11-JUL-1971  unknown          NaN           Cubana         1.0   
12199  12-JUL-1972  unknown          NaN              UTA         0.0   
13900  14-NOV-1978  unknown          NaN         Aeroflot         1.0   
14788  09-JUN-1982  unknown          NaN              LOT         0.0   
15224  25-JUN-1984  unknown          NaN             CAAC         0.0   
15851  05-MAY-1987  unknown          NaN         Iran Air         0.0   
16006  25-DEC-1987  unknown          NaN         Iran Air         0.0   
16016  05-JAN-1988  unknown          NaN         Iran Air         0.0   
16099  16-JUN-1988  unknown          NaN        Afghan AF        35.0   
16669  02-SEP-1990  unknown          NaN         Aeroflot         0.0   
16688  25-SEP-1990  unknown          NaN         Aeroflot         0.0   
16738  06-DEC-1990  unknown          NaN             CAAC         0.0   
16980  25-NOV-1991  unknown          NaN          Unknown         0.0   
17287  22-JAN-1993  unknown          NaN  Indian Airlines         0.0   
17479  22-OCT-1993  unknown          NaN         EgyptAir         0.0   
17662  26-MAY-1994  unknown          NaN           Garuda         0.0   
18137  24-MAR-1996  unknown          NaN    Sudan Airways         0.0   
18367  10-FEB-1997  unknown          NaN  China Northwest         0.0   
18993  26-OCT-1999  unknown          NaN         Iran Air         0.0   

                       location          country cat  year  
5446   near Xichang, Sichuan...            China  A1  1946  
5655        Shanghei-Lungwha...            China  A1  1947  
6178                     Tainan           Taiwan  H2  1949  
6537               near Aichach          Germany  H2  1950  
7036        Graz-Thalerhof A...          Austria  H2  1952  
8670        Bari-Palese Airp...            Italy  H2  1959  
8904                        NaN           Russia  H2  1960  
8972            near Cienfuegos             Cuba  H2  1960  
10170                       NaN  Unknown country  H2  1965  
10721                   Baghdad             Iraq  A1  1968  
11210       Managua Airport ...        Nicaragua  H2  1969  
11537       Warszawa-Okecie ...           Poland  H2  1970  
11636                       NaN  Unknown country  H2  1970  
11638                       NaN  Unknown country  H2  1970  
11711                      Brno       Czech Rep.  H2  1970  
11866       Havana-José Mar...             Cuba  H2  1971  
12199       Abidjan-Felix Ho...  Unknown country  H2  1972  
13900                       NaN  Unknown country  H2  1978  
14788       Warszawa-Okecie ...           Poland  H2  1982  
15224       Fuzhou Yixu Airp...            China  H2  1984  
15851       Tehran-Mehrabad ...             Iran  H2  1987  
16006       Mashhad Airport ...             Iran  H2  1987  
16016       Tehran-Mehrabad ...             Iran  H2  1

There are only 28 unknown types, so we will drop them since te type of aircraft is very important to us

In [14]:

acc_df = acc_df[acc_df["type"] != "unknown"]

In [15]:
acc_df[acc_df["type"] == "unknown"]

Empty DataFrame
Columns: [date, type, registration, operator, fatalities, location, country, cat, year]
Index: []

Now let's deal with the country and location columns

In [16]:
#Lets calculate the percentage of rows missing location data
(23019/23967) * 100
# We can see we have 4% missing values

96.04456127174865

In [17]:
acc_df[acc_df["location"].isnull()]

date                      type registration            operator  \
0        date unk.            Antonov An-12B       T-1206       Indonesian AF   
1        date unk.            Antonov An-12B       T-1204       Indonesian AF   
2        date unk.            Antonov An-12B       T-1201       Indonesian AF   
5        date unk.           Antonov An-12BP   CCCP-12172           Soviet AF   
8        date unk.            Antonov An-24B     RA-47794          Russian AF   
...            ...                       ...          ...                 ...   
19791  15-APR-2003      Vickers 836 Viscount       9Q-CGL        Trans Intair   
19806  10-MAY-2003             Antonov An-2R     UK-68038  Uzbekistan Airways   
20002  02-MAY-2004           Yakovlev Yak-40    YV-1070CP      CIACA Airlines   
20077  11-SEP-2004  Lockheed C-130H Hercules          NaN                USAF   
20093  09-OCT-2004              Let L-410UVP       9XR-KL      El Amiron Plus   

       fatalities location          country cat     year  
0             NaN      NaN  Unknown country  U1  unknown  
1             NaN      NaN  Unknown country  U1  unknown  
2             NaN      NaN  Unknown country  U1  unknown  
5             NaN      NaN           Russia  U1  unknown  
8             0.0      NaN           Russia  A1  unknown  
...           ...      ...              ...  ..      ...  
19791         0.0      NaN       D.R. Congo  A1     2003  
19806         NaN      NaN  Unknown country  A2     2003  
20002         NaN      NaN        Guatemala  A1     2004  
20077         0.0      NaN             Iraq  C2     2004  
20093         0.0      NaN       D.R. Congo  A1     2004  

[943 rows x 9 columns]

In [18]:
#lets drop rows where  the location is NaN
acc_df = acc_df.dropna(subset=["location"])

In [19]:
acc_df[acc_df["location"].isnull()] # Confirm that they are dropped

Empty DataFrame
Columns: [date, type, registration, operator, fatalities, location, country, cat, year]
Index: []

Null locations have been dropped since they are only 4% of the dataset. Let's now deal with the country column. We can see that country  has no null values. It however has unknown values. If they are not that many we can remove them since filling them in with the mode In my view will skew the data

In [20]:
acc_df["country"].value_counts() 

USA          4371
Russia       1366
U.K.          834
Canada        825
India         666
             ... 
Tuvalu          1
Mauritius       1
Kosovo          1
Macau           1
Eswatini        1
Name: country, Length: 232, dtype: int64

In [21]:
#Save this to a temporary csv to have a look at the data
acc_df[acc_df["country"] == "Unknown country"].to_csv("./output/unknown_countries.csv")

When we look at the temporary csv it has alot of missing data in the location, fatalities and countries columns. We can remove them safelt in my view.

In [22]:
acc_df = acc_df[acc_df["country"] != "Unknown country"]

In [23]:
acc_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22701 entries, 3 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   date          22701 non-null  object 
 1   type          22701 non-null  object 
 2   registration  21283 non-null  object 
 3   operator      22697 non-null  object 
 4   fatalities    19449 non-null  float64
 5   location      22701 non-null  object 
 6   country       22701 non-null  object 
 7   cat           22701 non-null  object 
 8   year          22701 non-null  object 
dtypes: float64(1), object(8)
memory usage: 1.7+ MB


In [24]:
acc_df[acc_df["location"]=="unknown"].count()

date            65
type            65
registration    60
operator        65
fatalities      28
location        65
country         65
cat             65
year            65
dtype: int64

In [25]:
#lets drop the location 